src/util/validation-error-builder.ts

```diff
@@ -0,0 +1,97 @@
+import { JSORMBase } from "../model"
+import {
+  JsonapiResponseDoc,
+  JsonapiError,
+  JsonapiErrorMeta
+} from "../jsonapi-spec"
+import { ValidationErrors } from "../validation-errors"
+
+export class ValidationErrorBuilder<T extends JSORMBase> {
+  static apply<T extends JSORMBase>(
+    model: T,
+    payload: JsonapiResponseDoc
+  ): void {
+    const instance = new ValidationErrorBuilder(model, payload)
+    instance.apply()
+  }
+
+  model: T
+  payload: JsonapiResponseDoc
+
+  constructor(model: T, payload: JsonapiResponseDoc) {
+    this.model = model
+    this.payload = payload
+  }
+
+  apply() {
+    const errorsAccumulator: ValidationErrors<T> = {}
+
+    if (!this.payload.errors) {
+      return
+    }
+
+    this.payload.errors.forEach(err => {
+      const meta = err.meta
+      const metaRelationship = meta.relationship
+
+      if (metaRelationship) {
+        this._processRelationship(this.model, metaRelationship, err)
+      } else {
+        this._processResource(errorsAccumulator, meta, err)
+      }
+    })
+
+    this.model.errors = errorsAccumulator
+  }
+
+  private _processResource(
+    errorsAccumulator: ValidationErrors<T>,
+    meta: JsonapiErrorMeta,
+    error: JsonapiError
+  ) {
+    let attribute = this.model.klass.deserializeKey(meta.attribute)
+
+    errorsAccumulator[attribute] = {
```

Comment:
I think we should probably add `raw` here as well, since the spec is pretty flexible with errors and server implementations vary

Reply:
Good idea.